pgrx-pg-sys/src/submodules/elog.rs

```diff
@@ -120,14 +215,109 @@ macro_rules! debug5 {
 /// [PostgreSQL settings](https://www.postgresql.org/docs/current/runtime-config-logging.html) are configured.
+///
+/// ## Adding Details and Hints
+///
+/// Additionally, you can use specify a `detail` and/or `hint` to be included with the report.
+/// After the string literal to be used as the message and any formatting arguments, add a semicolon.
+/// Then assign a string literal to `detail` or `hint`, optionally followed by a list of
+/// comma separated formatting arguments.
+///
+/// ## Examples
+///
+/// ```rust,no_run
+/// use pgrx_pg_sys::debug4;
+///
+/// // just like `println!` and `format!`
+/// debug4!("a simple message");
+/// debug4!("or a formatted message: {:?}", "pgrx rocks!");
+/// // include details and hints
+/// debug4! {
+///     "add details if you want";
+///     detail = "...";
+/// }
+/// debug4! {
+///     "or a message with just a hint";
+///     hint = "...";
+/// }
+/// debug4! {
+///     "put it all together for {} {}", "great", "success!";
+///     detail = "extra {}", "info";
+///     hint = "{} helpful", "very";
+/// }
+/// ```
 #[macro_export]
 macro_rules! debug4 {
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(detail = $detail:literal $(, $detail_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG4,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(detail = alloc::format!("{}", core::format_args!($detail $(, $detail_args)*));)?
+            );
+        }
+    );
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(hint = $hint:literal $(, $hint_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG4,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(hint = alloc::format!("{}", core::format_args!($hint $(, $hint_args)*));)?
+            );
+        }
+    );
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(hint = $hint:literal $(, $hint_args:expr)*)?;
+        $(detail = $detail:literal $(, $detail_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG4,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(detail = alloc::format!("{}", core::format_args!($detail $(, $detail_args)*));)?
+                $(hint = alloc::format!("{}", core::format_args!($hint $(, $hint_args)*));)?
```

Comment:
All of this eager allocation is starting to make me slightly unhappy. Perhaps it's unimportant in the grand scheme of things, and I don't have time to look into the impl deep enough to figure it out right now, but I really hope we aren't just allocating here only to palloc and copy it later? I won't block this change on that, but I would like to know.

Reply:
I don't think there is any extra allocation that could otherwise be avoided. [`set_detail` and `set_hint`](https://github.com/pgcentralfoundation/pgrx/blob/a5625169297cd378c3d206cbf501025b5b2f6ba8/pgrx-pg-sys/src/submodules/panic.rs#L266C1-L276C6) have the trait bound `S: Into<String>` so the resulting owned `String` from `format` should pass through without being copied. 

If my logic is correct then the original cases should be updated to omit the `.as_str()` to avoid an extra copy.

> Original implementation
```rust
$crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION, alloc::format!($($arg)*).as_str());
```

Reply:
That's correct, but we control all the signatures of everything so we can change those signatures if we want.

Reply:
After looking deeper into the `.report()` implementation I see your point, there is potential to eliminate all Rust allocation in the case where it is known ahead of time that the payload will be handed off to `palloc` immediately. 

My first thoughts on how to achieve this would be:
- Define a new type like ErrorReportWithLevel with fields of fmt::Arguments rather than String.
- Define a new fn like .as_pg_cstr that will facilitate copying to palloc pointer via the Write trait.

^ What are your first impressions? If this seems reasonable enough I can put something together _(I might need some help/guidance on creating unit tests for it. I admittedly have limited experience writing tests for unsafe code)_

Reply:
I see, that is what I expected. I think that can wait until a later PR since it's _technically_ unrelated. I think I would avoid using quite so many indirections, also, and I can explain how slightly better in another context.